src/lib.rs

```diff
@@ -78,23 +92,39 @@ pin_project! {
 /// Builder for `TlsListener`.
 #[derive(Clone)]
 pub struct Builder {
-    server_config: Arc<ServerConfig>,
+    acceptor: TlsAcceptor,
     max_handshakes: usize,
     handshake_timeout: Duration,
 }
 
+/// Wraps errors from either the listener or the TLS Acceptor
+#[derive(Debug, Error)]
+pub enum Error<A: std::error::Error> {
+    /// An error that arose from the listener ([AsyncAccept::Error])
+    #[error("{0}")]
+    ListenerError(#[source] A),
+    /// An error that occurred during the TLS accept handshake
+    #[cfg(feature = "rustls")]
+    #[error("{0}")]
+    TlsAcceptError(#[source] std::io::Error),
+    /// An error that occurred during the TLS accept handshake
+    #[cfg(feature = "native-tls")]
+    #[error("{0}")]
+    TlsAcceptError(#[source] tokio_native_tls::native_tls::Error),
+}
+
```

Comment:
Before we were relying on the fact that all implementations of AsyncAccept happened to return io::Error, and also that the TLS accept future also returned an io::Error, but we can't do that anymore.

This also gets rid of the slightly odd downcasting from below and replaces the error with a type parameter